linux/gatt/client.go

```diff
@@ -431,16 +440,22 @@ func (p *Client) HandleNotification(req []byte) {
 	sub, ok := p.subs[vh]
 	if !ok {
 		// FIXME: disconnects and propagate an error to the user.
-		log.Printf("Got an unregistered notification")
+		log.Printf("got an unregistered notification")
 		return
 	}
-	fn := sub.nHandler
-	if req[0] == att.HandleValueIndicationCode {
-		fn = sub.iHandler
-	}
-	if fn != nil {
-		fn(req[3:])
+
+	indication := req[0] == att.HandleValueIndicationCode
+	nd := req[3:]
+
+	switch {
+	case indication && sub.iHandler != nil:
+		sub.iHandler(sub.id, nd)
+	case sub.nHandler != nil:
+		sub.nHandler(sub.id, nd)
+	default:
+		log.Printf("no handler, dropping data vh 0x%x, indication %v, id %v, %v", vh, indication, sub.id, hex.EncodeToString(nd))
 	}
+	sub.id++
```

Comment:
Is there maybe a race here where a new notification comes in before this one gets processed thereby having the same id? Maybe the id should be incremented before the notification is sent to the handler func?

Reply:
the entire `HandleNotification` function is run under lock so this is ok.